src/System.IO.Pipelines/PipelineConnectionExtensions.cs

```diff
@@ -19,16 +19,30 @@ public static Stream GetStream(this IPipeConnection connection)
 
     public static class PipelineWriterExtensions
     {
+        private readonly static Task _completedTask = Task.FromResult(0);
+
         public static Task WriteAsync(this IPipeWriter output, Span<byte> source)
         {
             var writeBuffer = output.Alloc();
             writeBuffer.Write(source);
-            return AwaitFlushAsync(writeBuffer);
+            return FlushAsync(writeBuffer);
+        }
+
+        private static Task FlushAsync(WritableBuffer writeBuffer)
+        {
+            var awaitable = writeBuffer.FlushAsync();
+            if (awaitable.IsCompleted)
+            {
+                awaitable.GetResult();
```

Comment:
Is also `WritableBufferAwaitable` is custom awaiter on top of a [`IWritableBufferAwaiter` interface](https://github.com/dotnet/corefxlab/blob/master/src/System.IO.Pipelines/IReadableBufferAwaiter.cs#L17-L24)  rather than a Task so can make less assumptions on how it behaves.

So for example it could be a reusable auto-reseting type; that would become incomplete once it completed/returned on an await; so calling `GetResult()` maintains the normal flow whatever that is.